### PR TITLE
Support projects hosted in locations other than GitHub

### LINKED ISSUE #11: Support projects hosted in locations other than GitHub
Various properties across the files, specifically in:
- Endjin.Common.GitHub.props
- Endjin.Common.NuGet.props
- Endjin.RecommendedPractices.nuspec

assume that the project is being hosted in GitHub. When trying to use the package with a solution hosted elsewhere - e.g. Azure DevOps, various warnings are displayed.

### PR DESCRIPTION
**Pull request by idg10**
Resolves #11

### CODE REVIEW

**Comment by HowardvanRooijen** [comment]
Is the significant challenge really just going through all the project and manually updating them to use the required packages?

**Comment by idg10** [comment]
>Is the significant challenge really just going through all the project and manually updating them to use the required packages?

The hard part of this is "all the projects". This is a public NuGet package. We've encouraged customers to use it. We've also written at least one blog encouraging others to use it. So we have no way to discover all the places it is used.

The dependency analysis in GitHub Insights only shows our projects using it, but who knows what non-public projects might be using it?

So my view is that if someone upgrades from v1 to v2, and they were previously relying on the GitHub integration, we shouldn't just switch that off without telling them.

Right now I'm looking into whether it's possible for us to detect that we're in a git repo that is on GitHub and to produce a message along the lines of "Hey, it looks like you're using GitHub—if you just upgraded from v1, you might now want to use `Endjin.RecommendedPractices.GitHub` instead, but if not set this property to silence this warning".

So that way, in the single most common expected scenario—upgrading a GitHub-hosted project from v1 to v2—people will be told what they need to do. And in the two new scenarios that this supports, we won't produce a spurious warning (assuming I can in fact get detection of being hosted on GitHub to work). So the only situation in which someone will see this warning is if they are hosting on GitHub, they do want `Endjin.RecommendedPractice` but they don't want sourcelink. (And they'll have to add a setting saying that this is deliberate to silence the warning.)

My current view is that this would be better than silently losing functionality in a way that won't be initially very obvious, which is the alternative. (In particular, sourcelink will break, and people probably won't notice that right away.)

**Comment by HowardvanRooijen** [comment]
Could we do a point release with a depreciation warning? Or add a warning when building in a GH repo that you no longer have the GitHub package installed?

**Comment by idg10** [comment]
@HowardvanRooijen wrote:

>...Or add a warning when building in a GH repo that you no longer have the GitHub package installed?

Yes. I've now done this.

If you install just `Endjin.RecommendedPractices` (and not `Endjin.RecommendedPractices.GitHub`) in a solution that is in fact in GitHub, we now detect this, and produce a warning suggesting that maybe you want the `Endjin.RecommendedPractices.GitHub` package instead (along with details on how to disable that warning in case it really was deliberate).

This also works for anyone who upgrades from v1 to v2 will see this warning telling them they probably want the new package to get GitHub Source Link support.

And if you install `Endjin.RecommendedPractices` v2 in a non-GitHub-hosted repo, you get all the functionality except for the GitHub Source Link support, and no warnings.

**Comment by HowardvanRooijen** [comment]
And is there the ability to suppress the warning via a property in case I really don't want GitHub integration in a repo and want to mute the warning?

**Comment by idg10** [comment]
> And is there the ability to suppress the warning via a property in case I really don't want GitHub integration in a repo and want to mute the warning?

Yes. And the warning tells you which property to set.

**Comment by HowardvanRooijen** [comment]
🔥🎉👌

**Review by mwadams — APPROVED** [review]
Very nice! It is sad that we can't suppress the warning except with a hacky conditional that essentially says "if this looks like a git repo, check to see if this looks like a git repo".

File: src/Endjin.RecommendedPractices.NuGet/build/Endjin.IssueGitHubWarningIfAppropriate.targets [hunk 1]
```diff
@@ -0,0 +1,75 @@
+<!--
+    In V1, we always added an implicit reference to the GitHub sourcelink components. As of V2, we
+    support use in projects that are not in GitHub. However, we anticipate that most V1->V2
+    migrations will want to keep GitHub sourcelink, so we want to make it hard to lose that by
+    accident. Developers will need to change their reference from Enddin.RecommendedPractices to
```
Comment (HowardvanRooijen) [inline src/Endjin.RecommendedPractices.NuGet/build/Endjin.IssueGitHubWarningIfAppropriate.targets:5]: Typo "Enddin"

**Comment by idg10** [comment]
@mwadams wrote:

> Very nice! It is sad that we can't suppress the warning except with a hacky conditional that essentially says "if this looks like a git repo, check to see if this looks like a git repo".

Yes. I've created this outlining a couple of alternatives: https://github.com/endjin/Endjin.RecommendedPractices.NuGet/issues/33

I don't think it's urgent though. As long as the hack is working for all, it can stay, because a better fix like that wouldn't require any major changes in usage model or design.

**Pull request merged**